src/ExtensibleService.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Di;
+
+use Psr\Container\ContainerInterface;
+use Yiisoft\Factory\Definition\DefinitionInterface;
+use Yiisoft\Factory\Definition\Normalizer;
+use Yiisoft\Factory\DependencyResolverInterface;
+
+/**
+ * A wrapper for a service definition that allows registering extensions.
+ * An extension is a callable that returns a modified service object:
+ *
+ * ```php
+ * static function (ContainerInterface $container, $service) {
+ *     return $service->withAnotherOption(42);
+ * }
+ * ```
+ */
+final class ExtensibleService implements DefinitionInterface
+{
+    /** @psalm-var  array<string,mixed> */
+    private $definition;
+    private array $extensions = [];
+
+    /**
+     * @param mixed $definition Definition to allow registering extensions for.
+     */
+    public function __construct($definition)
+    {
+        $this->definition = $definition;
+    }
+
+    /**
+     * Add an extension.
+     *
+     * An extension is a callable that returns a modified service object:
+     *
+     * ```php
+     * static function (ContainerInterface $container, $service) {
+     *     return $service->withAnotherOption(42);
+     * }
+     * ```
+     *
+     * @param \Closure $closure An extension to register.
+     */
+    public function addExtension(\Closure $closure): void
```

Comment:
What do you think about making it a `callable`?

```suggestion
    public function addExtension(callable $closure): void
```